README.md
# JAWS - Just another WiFi scanner
A simple free and open source wifi scanner that supports real time scans of nearby networks. It displays a list of nearby wifi networks ordered by signal strength and periodically rescans for new networks.

[![Google Play](https://developer.android.com/images/brand/en_generic_rgb_wo_60.png)](https://play.google.com/store/apps/details?id=is.pinterjann.jaws) [![F-Droid](https://f-droid.org/wiki/images/1/13/F-Droid-button.svg)](https://f-droid.org/repository/browse/?fdid=is.pinterjann.jaws)

![screenshots](https://raw.githubusercontent.com/jannispinter/jaws/master/screenshots.png)

# Features
- Display a list of nearby wifi networks ordered by signal strength
- Display SSID, BSSID, Channel, Capabilities and the signal strength of each network
- Periodically rescan for new networks

# Goals and motivation
- Create a free and open source wifi scanner that is available on F-Droid
- Provide a feature rich but yet beautiful, easy to use UI
- Implement useful, optional features such as a war driving mode without bloating the UI

# To do list
- Group networks by SSID
- "Find network" feature
- Signal strength diagram (e.g. last 60 seconds)
- Improve UX

# License
 ![gpl3](https://www.gnu.org/graphics/gplv3-127x51.png)

  JAWS is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  JAWS is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.


CHANGELOG.md
###Changelog

####Version 0.5
* Fixed Android 6 permission issue

####Version 0.4
* Improved battery saving feature
* Fixed network ordering preference default value
* Fixed a typo (en) in preferences
* Added a feature to pin favorite networks to the top
* Improved Hebrew translation (thanks GreenLunar)

####Version 0.3
* Improved UX and UI
* Implemented preferences
* Scan interval can now be adjusted in preferences
* Battery saving option implemented (turn off WiFi after scanning)

####Version 0.2.1
* Added Hebrew translation (thanks GreenLunar)
* Fixed F-Droid build

####Version 0.2
* Removed dynamic signal color generation from donut circle
* Replaced dynamic signal color generation with static pastel colors
* Added new capability badges

####Version 0.1
* initial release 


app/src/main/java/is/pinterjann/jaws/helper/SignalColor.java
package is.pinterjann.jaws.helper;

import android.graphics.Color;
import android.net.wifi.WifiManager;

public class SignalColor {

    public static int BOTTLE_GREEN      = Color.rgb(0, 106, 78);
    public static int PASTEL_ORANGE     = Color.rgb(255,179,71);
    public static int PASTEL_RED        = Color.rgb(255, 105, 97);
    public static int PASTEL_RED_DARK   = Color.rgb(194, 59, 34);

    public static int getColor(int signal) {

        /* in case the network is out of range */
        if(signal == 0) {
            return PASTEL_RED_DARK;
        }

        int signalPercent = WifiManager.calculateSignalLevel(signal, 5);

        /* carefully chosen pastel colors */
        switch (signalPercent) {
            case 3:  return BOTTLE_GREEN;
            case 2:  return PASTEL_ORANGE;
            case 1:  return PASTEL_RED;
            case 0:  return PASTEL_RED_DARK;
            default: return BOTTLE_GREEN;
        }
    }
}


app/src/main/java/is/pinterjann/jaws/model/WirelessNetwork.java
package is.pinterjann.jaws.model;

import java.util.Date;

public class WirelessNetwork implements Comparable {
    private String bssid;
    private String ssid;
    private int channel;
    private int signal;
    private String security;
    private long timestamp;

    public WirelessNetwork(String bssid, String ssid, int channel, int signal, String security, long timestamp) {
        this.bssid = bssid;
        this.ssid = ssid;
        this.channel = channel;
        this.signal = signal;
        this.security = security;
        this.timestamp = timestamp;
    }

    /* Builds a network from a given string produced by WirelessNetwork.toString() */
    public WirelessNetwork(String networkString) {

        // TODO: add some input validation

        String[] networkInfo = networkString.split(";");
        this.bssid = networkInfo[0];
        this.channel = Integer.parseInt(networkInfo[1]);
        this.security = networkInfo[2];
        this.ssid = networkInfo[3];

        this.signal = 0;
        this.timestamp = new Date().getTime();
    }

    public String getBssid() {
        return bssid;
    }

    public void setBssid(String bssid) {
        this.bssid = bssid;
    }

    public String getSsid() {
        return ssid;
    }

    public void setSsid(String ssid) {
        this.ssid = ssid;
    }

    public int getChannel() {
        return channel;
    }

    public void setChannel(int channel) {
        this.channel = channel;
    }

    public int getSignal() {
        return signal;
    }

    public void setSignal(int signal) {
        this.signal = signal;
    }

    public String getSecurity() {
        return security;
    }

    public void setSecurity(String security) {
        this.security = security;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    @Override
    public boolean equals(Object otherNetwork) {
        if(otherNetwork instanceof  WirelessNetwork) {
            WirelessNetwork network = (WirelessNetwork) otherNetwork;
            if (getBssid().equals(network.getBssid()) &&
                    getSsid().equals(network.getSsid()) &&
                    getChannel() == network.getChannel() &&
                    getSecurity().equals(network.getSecurity())) {
                return true;
            }
        }
        return false;
    }

    @Override
    public int compareTo(Object otherNetwork) {
        return ((WirelessNetwork)otherNetwork).getSignal() - this.getSignal();
    }

    @Override
    public String toString() {
        return getBssid() + ";" + getChannel() + ";" + getSecurity() + ";" + getSsid();
    }

}


app/src/main/java/is/pinterjann/jaws/adapter/NetworkAdapter.java
package is.pinterjann.jaws.adapter;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.wifi.WifiManager;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import com.github.lzyzsd.circleprogress.DonutProgress;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import is.pinterjann.jaws.helper.SignalColor;
import is.pinterjann.jaws.model.WirelessNetwork;
import is.pinterjann.jaws.R;

public class NetworkAdapter extends BaseAdapter {

    private List<WirelessNetwork> networkList = new ArrayList<>();
    private SharedPreferences sharedPreferences;
    private Set<String> pinnedNetworks;

    public List<WirelessNetwork> getNetworkList() {
        return networkList;
    }

    public void setNetworkList(List<WirelessNetwork> networkList) {
        this.networkList = networkList;
        notifyDataSetChanged();
    }

    @Override
    public int getCount() {
        return networkList.size();
    }

    @Override
    public Object getItem(int position) {
        return networkList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(parent.getContext());
        pinnedNetworks = sharedPreferences.getStringSet("pinned_networks", new HashSet<String>());

        LayoutInflater inflater = (LayoutInflater)
                parent.getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);

        if(convertView == null) {
            convertView = inflater.inflate(R.layout.network_row, null);
        }

        TextView ssid               = (TextView) convertView.findViewById(R.id.network_ssid);
        TextView bssid              = (TextView) convertView.findViewById(R.id.network_bssid);
        TextView signal             = (TextView) convertView.findViewById(R.id.network_signal);
        TextView channel            = (TextView) convertView.findViewById(R.id.network_channel);
        ImageView cap_badge_ess     = (ImageView) convertView.findViewById(R.id.cap_badge_ess);
        ImageView cap_badge_crypto  = (ImageView) convertView.findViewById(R.id.cap_badge_crypto);
        ImageView cap_badge_wps     = (ImageView) convertView.findViewById(R.id.cap_badge_wps);
        final ImageButton btn_stick       = (ImageButton) convertView.findViewById(R.id.network_btn_stick);
        DonutProgress signalDonutProgress   = (DonutProgress)
                convertView.findViewById(R.id.network_donut_progress);


        final WirelessNetwork network = networkList.get(position);
        ssid.setText(network.getSsid());
        bssid.setText(network.getBssid());
        signal.setText(network.getSignal() + " dBm");
        signal.setTextColor(SignalColor.getColor(network.getSignal()));
        String channelTranslation = convertView.getResources().getString(R.string.network_channel);
        channel.setText(channelTranslation + ": " + network.getChannel());

        /* Set donut circle signal strength */
        if(network.getSignal() != 0) {
            signalDonutProgress.setProgress(WifiManager.calculateSignalLevel(network.getSignal(), 100) + 1);
            signalDonutProgress.setTextColor(signalDonutProgress.getFinishedStrokeColor());
        } else {
            signalDonutProgress.setProgress(0);
            signalDonutProgress.setTextColor(signalDonutProgress.getUnfinishedStrokeColor());
        }

        /* Check ESS */
        if(network.getSecurity().contains("ESS")) {
            cap_badge_ess.setVisibility(View.VISIBLE);
        } else {
            cap_badge_ess.setVisibility(View.INVISIBLE);
        }

        /* Check cryptography */
        if(network.getSecurity().contains("WPA2-")) {
            cap_badge_crypto.setImageResource(R.mipmap.cap_badge_wpa2);
        } else if (network.getSecurity().contains("WPA-")) {
            cap_badge_crypto.setImageResource(R.mipmap.cap_badge_wpa);
        } else if(network.getSecurity().contains("WEP")) {
            cap_badge_crypto.setImageResource(R.mipmap.cap_badge_wep);
        } else {
            cap_badge_crypto.setImageResource(R.mipmap.cap_badge_open);
        }

        /* Check WPS */
        if(network.getSecurity().contains("WPS")) {
            cap_badge_wps.setVisibility(View.VISIBLE);
        } else {
            cap_badge_wps.setVisibility(View.INVISIBLE);
        }

        /* Update pinned/unpinned graphic */
        if(pinnedNetworks.contains(network.toString())) {
            btn_stick.setImageResource(R.drawable.ic_pin_pinned);
        } else {
            btn_stick.setImageResource(R.drawable.ic_pin_unpinned);
        }


        btn_stick.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                SharedPreferences.Editor editor = sharedPreferences.edit();

                // Due to a bug in Android, we need to make a copy of the set. Otherwise, the changes
                // to the set are not persistent.
                // See: https://code.google.com/p/android/issues/detail?id=27801
                Set<String> pinnedNetworksSet = new HashSet<>();
                pinnedNetworksSet.addAll(pinnedNetworks);


                if(!pinnedNetworksSet.contains(network.toString())) {
                    pinnedNetworksSet.add(network.toString());
                    editor.putStringSet("pinned_networks", pinnedNetworksSet);
                    editor.commit();
                    btn_stick.setImageResource(R.drawable.ic_pin_pinned);
                } else {
                    pinnedNetworksSet.remove(network.toString());
                    editor.putStringSet("pinned_networks", pinnedNetworksSet);
                    editor.commit();
                    btn_stick.setImageResource(R.drawable.ic_pin_unpinned);
                }

                notifyDataSetChanged();
            }
        });

        return convertView;
    }

}


app/src/main/java/is/pinterjann/jaws/activities/AboutActivity.java
package is.pinterjann.jaws.activities;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import is.pinterjann.jaws.R;

public class AboutActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_about);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                this.finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }
}


app/src/main/java/is/pinterjann/jaws/helper/FrequencyConverter.java
package is.pinterjann.jaws.helper;

public class FrequencyConverter {

    /**
     * Takes a frequency in MHz and converts it to the corresponding WiFi channel number
     * @param frequency
     * @return channel
     */
    public static int convert(int frequency) {

        switch(frequency) {
            /* 2.4 GHz Band */
            case 2412: return 1;
            case 2417: return 2;
            case 2422: return 3;
            case 2427: return 4;
            case 2432: return 5;
            case 2437: return 6;
            case 2442: return 7;
            case 2447: return 8;
            case 2452: return 9;
            case 2457: return 10;
            case 2462: return 11;
            case 2467: return 12;
            case 2472: return 13;
            case 2484: return 14;

            /* 5 GHz Band */
            case 5180: return 36;
            case 5200: return 40;
            case 5220: return 44;
            case 5240: return 48;
            case 5260: return 52;
            case 5280: return 56;
            case 5300: return 60;
            case 5320: return 64;
            case 5500: return 100;
            case 5520: return 104;
            case 5540: return 108;
            case 5560: return 112;
            case 5580: return 116;
            case 5600: return 120;
            case 5620: return 124;
            case 5640: return 128;
            case 5660: return 132;
            case 5680: return 136;
            case 5700: return 140;
            case 5745: return 149;
            case 5765: return 153;
            case 5785: return 157;
            case 5805: return 161;
            case 5825: return 165;

            /* Other frequencies are not implemented yet */
            default: return 0;
        }
    }

}


app/src/main/java/is/pinterjann/jaws/activities/JAWSActivity.java
package is.pinterjann.jaws.activities;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.net.wifi.ScanResult;
import android.net.wifi.WifiManager;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ListView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import is.pinterjann.jaws.R;
import is.pinterjann.jaws.adapter.NetworkAdapter;
import is.pinterjann.jaws.helper.FrequencyConverter;
import is.pinterjann.jaws.model.WirelessNetwork;

public class JAWSActivity extends AppCompatActivity {

    private WifiManager wifiManager;
    private NetworkAdapter networkAdapter;
    private BroadcastReceiver wifiScanReceiver;
    private SharedPreferences sharedPreferences;
    private Toolbar toolbar;

    private boolean isScanning = true;
    private boolean jawsAutoEnabledWifi = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_jaws);

        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
        toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        ListView listview = (ListView) findViewById(R.id.list);
        networkAdapter = new NetworkAdapter();
        listview.setAdapter(networkAdapter);

        wifiManager = (WifiManager) getSystemService(Context.WIFI_SERVICE);

        /* Fake networks for screenshots
        List<WirelessNetwork> networkList = new ArrayList<>();
        networkList.add(new WirelessNetwork("6E:AD:9E:53:02:B4", "Guest", 1, -80, "ESS OPEN", new Date().getTime()));
        networkList.add(new WirelessNetwork("54:4F:7C:33:6B:4F", "Telcom", 5, -77, "ESS OPEN", new Date().getTime()));
        networkList.add(new WirelessNetwork("CB:1A:07:DA:2D:EC", "Freifunk", 9, -58, "ESS OPEN", new Date().getTime()));
        networkList.add(new WirelessNetwork("60:B0:56:A7:95:07", "Free Manning",13, -57, "ESS WPA", new Date().getTime()));
        networkList.add(new WirelessNetwork("52:22:1E:E4:62:29", "Freedom", 100, -90, "ESS OPEN", new Date().getTime()));
        networkList.add(new WirelessNetwork("F8:19:B0:58:CA:96", "private", 36, -67, "ESS WPA2", new Date().getTime()));
        networkList.add(new WirelessNetwork("9D:44:E7:25:7A:CA", "JAWS 1", 40, -60, "ESS WEP", new Date().getTime()));
        networkList.add(new WirelessNetwork("69:40:1E:93:1F:D3", "JAWS 2", 120, -72, "ESS WPA", new Date().getTime()));
        networkList.add(new WirelessNetwork("7F:6D:12:88:D7:97", "JAWS 3", 1, -63, "ESS WPA2", new Date().getTime()));
        networkList.add(new WirelessNetwork("49:97:33:B9:25:4E", "JAWS 4", 5, -59, "ESS WPA2 WPS", new Date().getTime()));
        Collections.sort(networkList);
        networkAdapter.setNetworkList(networkList);
        findViewById(R.id.loadingPanel).setVisibility(View.GONE);
        */

        wifiScanReceiver = new WifiScanReceiver();
        registerReceiver(wifiScanReceiver, new IntentFilter(
                WifiManager.SCAN_RESULTS_AVAILABLE_ACTION));

        startScanning();
    }

    @Override
    protected void onPause() {
        super.onPause();
        stopScanning();
    }

    @Override
    protected void onResume() {
        super.onResume();
        startScanning();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unregisterReceiver(wifiScanReceiver);
        stopScanning();

        if(sharedPreferences.getBoolean("switch_wifi_off_when_not_scanning", false) && jawsAutoEnabledWifi) {
            wifiManager.setWifiEnabled(false);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_jaw, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.action_settings) {
            Intent intent = new Intent(this, SettingsActivity.class);
            startActivity(intent);
            return true;
        } else if (id == R.id.action_about) {
            Intent intent = new Intent(this, AboutActivity.class);
            startActivity(intent);
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private void startScanning() {
        isScanning = true;

        if (!wifiManager.isWifiEnabled()) {
            jawsAutoEnabledWifi = true;
            String toastTranslation = getResources().getString(R.string.toast_wifi_enabled);
            Toast.makeText(getApplicationContext(), toastTranslation,
                    Toast.LENGTH_LONG).show();
            wifiManager.setWifiEnabled(true);
        }
        /* Start AsyncTask to scan for networks in the background */
        new WifiScanAsyncTask().execute();
    }

    private void stopScanning() {
        isScanning = false;
    }

    public class WifiScanReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context c, Intent intent) {
            List<ScanResult> results = wifiManager.getScanResults();
            List<WirelessNetwork> networkList = new ArrayList<>();
            for (android.net.wifi.ScanResult result : results) {
                WirelessNetwork network = new WirelessNetwork(result.BSSID, result.SSID,
                        FrequencyConverter.convert(result.frequency), result.level,
                        result.capabilities, new Date().getTime()
                );
                networkList.add(network);
            }

            /* Sort networks based on preferences */
            if (sharedPreferences.getString("network_order", "first").equals("first")) {
                Collections.sort(networkList);
            } else {
                Collections.sort(networkList, new Comparator<WirelessNetwork>() {
                    @Override
                    public int compare(WirelessNetwork lhs, WirelessNetwork rhs) {
                        return lhs.getSignal() - rhs.getSignal();
                    }
                });
            }

            /* Add pinned networks to the top of the list */
            Set<String> pinnedNetworks = sharedPreferences.getStringSet("pinned_networks", new HashSet<String>());
            for (String networkString : pinnedNetworks) {
                WirelessNetwork network = new WirelessNetwork(networkString);
                if(networkList.contains(network)) {
                    int position = networkList.indexOf(network);
                    WirelessNetwork origNetwork = networkList.get(position);
                    networkList.remove(position);
                    networkList.add(0, origNetwork);
                } else {
                    networkList.add(0, network);
                }
            }

            /* Disable loading animation */
            findViewById(R.id.loadingPanel).setVisibility(View.GONE);

            networkAdapter.setNetworkList(networkList);
        }
    }

    public class WifiScanAsyncTask extends AsyncTask<Void, Void, Void> {

        @Override
        protected Void doInBackground(Void... params) {

            while(isScanning) {
                wifiManager.startScan();

                try {
                    int interval = Integer.parseInt(sharedPreferences.getString("scan_interval", "1000"));
                    Thread.sleep(interval);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

            return null;
        }
    }
}
